### PR TITLE
Skip default delete

### PR DESCRIPTION
**Pull request by danielrbradley**
For resources without DELETE operations defined (only PUT) and no obvious "default" state to reset to, add the option for delete to be a no-op. 

This is required for Sql database TransparentDataEncryption which requires different properties to be set for different API versions. This could be solved by introducing version awareness, but that would be a larger piece of work.

We model skipping the delete by setting the "default body" to nil.

This includes a refactor to disentwine path normalisation and default resource state from the main discovery process. These are moved into their own modules so only public members can be interacted with - to avoid interacting with the wrong internal parts.

### CODE REVIEW

File: provider/pkg/provider/provider.go [hunk 1]
```diff
@@ -1260,8 +1260,8 @@ func (k *azureNativeProvider) Delete(ctx context.Context, req *rpc.DeleteRequest
 	case res.Singleton:
 		// Singleton resources can't be deleted (or created), set them to the default state.
 		for _, param := range res.PutParameters {
-			if param.Location == "body" && res.DefaultBody != nil {
-				requestBody := k.converter.SdkPropertiesToRequestBody(param.Body.Properties, *res.DefaultBody)
+			if param.Location == "body" && res.SkipDelete {
```
Comment (thomas11) [inline provider/pkg/provider/provider.go:1263]: The comment for `skipDelete` says 
> true for resources that don't have a delete operation and _no sensible default_ to set its state back to

but here we set the default if the property is true?

Comment (danielrbradley) [reply provider/pkg/provider/provider.go:1263]: Oops yup that's missing a `!`

File: provider/pkg/openapi/defaults/defaultResourcesState.go [hunk 2]
```diff
@@ -179,3 +204,15 @@ func containsNonEmptyCollections(value map[string]interface{}) bool {
 	}
 	return false
 }
+
+type DefaultResourceState struct {
+	HasDefault             bool
```
Comment (thomas11) [inline provider/pkg/openapi/defaults/defaultResourcesState.go:209]: A default state that has no default (`!HasDefault`) reads a bit odd. Maybe `HasDefaultBody` or `HasDefaultProperties`?

**Review by thomas11 — APPROVED** [review]
LGTM, although I'd be more confident with some test coverage.

File: provider/pkg/openapi/defaults/defaultResourcesState.go (hunk 2 above)
Comment (danielrbradley) [reply provider/pkg/openapi/defaults/defaultResourcesState.go:209]: Hmm, this was to indicate if we actually have any form of delete behaviour specified in the map. Not sure on the most "ideomatic go" here ... a second return for "ok" or rename this to `Found` perhaps ... or just make the whole return a pointer and return nil if not found 🤔

Comment (thomas11) [reply provider/pkg/openapi/defaults/defaultResourcesState.go]: I think the structure is ok, I'd just consider a different name for `HasDefault`.

File: provider/pkg/openapi/defaults/defaultResourcesState_test.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+func TestNonEmptyCollection(t *testing.T) {
+	actual := GetDefaultResourceState("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Insights/components/{resourceName}/currentbillingfeatures")
+	expected := DefaultResourceState{
+		State: map[string]interface{}{
+			"currentBillingFeatures": []string{"Basic"},
+			"dataVolumeCap":          map[string]string{},
+		},
+		HasNonEmptyCollections: true,
+	}
+	assert.Equal(t, &expected, actual)
```
Comment (mikhailshilkov) [inline provider/pkg/openapi/defaults/defaultResourcesState_test.go:49]: Could you please explain what we are testing here? `GetDefaultResourceState` does two things: 

- Calls `NormalizePath`
- Calls a map lookup and returns the value or default

What do we get by having 5 tests for it? It feels like we are testing the contents of `defaultResourcesStateNormalized`?

**Pull request merged**